core/ledger/ledgermgmt/ledger_mgmt.go

```diff
@@ -112,10 +122,35 @@ func (m *LedgerMgr) CreateLedger(id string, genesisBlock *common.Block) (ledger.
 	}, nil
 }
 
-// CreateLedgerFromSnapshot creates a new ledger with the given snapshot and returns the ledger and channel id.
-// This function guarantees that the creation of ledger and all ledger dbs would be an atomic action.
-// The channel id retrieved from the snapshot metadata is treated as a ledger id
-func (m *LedgerMgr) CreateLedgerFromSnapshot(snapshotDir string) (ledger.PeerLedger, string, error) {
+// CreateLedgerFromSnapshot creates a new ledger with the given snapshot and executes the callback function after the ledger is created.
+// This function creates the ledger and calls the callback func in a separate goroutine all ledger dbs would be an atomic action.
+// The channel id retrieved from the snapshot metadata is treated as a ledger id.
+// It returns an error if another ledger is being created from a snapshot.
+func (m *LedgerMgr) CreateLedgerFromSnapshot(snapshotDir string, channelCallback func(ledger.PeerLedger, string)) error {
+	if err := m.setUnderCreationLedger(snapshotDir); err != nil {
+		return err
+	}
+
+	go func() {
+		defer m.unsetUnderCreationLedger()
+
+		ledger, cid, err := m.createFromSnapshot(snapshotDir)
+		if err != nil {
+			logger.Errorw("Error creating ledger from snapshot", "snapshotDir", snapshotDir, "error", err)
+			return
+		}
+		if cid != m.underCreationLedger {
+			// this should not happen unless there is a code bug
+			logger.Errorw("Mismatched ledger ID", "createdLedger", cid, "underCreationLedger", m.underCreationLedger)
+		}
+
+		channelCallback(ledger, cid)
+	}()
+
+	return nil
+}
+
+func (m *LedgerMgr) createFromSnapshot(snapshotDir string) (ledger.PeerLedger, string, error) {
 	m.lock.Lock()
 	defer m.lock.Unlock()
```

Comment:
Now, is this lock needed?.. rather this begs a question whether we really need two locks at all? - If you release this lock here after setting the underconstruction status.

Reply:
Good point. One lock should be sufficient.

Reply:
After discussion with Manish and Senthil, we decided to keep the lock so that it does not change the current semantics where ledger mgr only allows one operation at a time.